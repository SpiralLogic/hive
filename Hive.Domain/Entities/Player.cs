using System.Collections.Generic;

namespace Hive.Domain.Entities
{
    public class Player
    {
        public Player(int id, string name, ICollection<Tile> availableTiles = null)
        {
            Id = id;
            Name = name;
            AvailableTiles = availableTiles ?? new List<Tile>();
        }

        public int Id { get; }
        public string Name { get; }

        public ICollection<Tile> AvailableTiles { get; }
    }
}