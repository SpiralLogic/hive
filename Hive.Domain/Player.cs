using System;
using System.Collections.Generic;
using System.Drawing;

namespace Hive.Domain
{
    public class Player
    {
        private static int _count = 0;
        
        public Player(string name, string color, string tileListColor)
        {
            Id = (++_count).ToString();
            Name = name;
            Color = color;
            TileListColor = tileListColor;
        }

        public string Id { get; }
        public string Name { get; }
        public string Color { get; }
        public string TileListColor { get; }
        
        public ICollection<Tile> AvailableTiles { get; } = new List<Tile>();
    }
}