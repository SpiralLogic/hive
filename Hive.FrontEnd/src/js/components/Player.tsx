import * as React from 'react';
import { Tile } from './Tile';
import { IPlayer } from '../domain';

export const Player: React.FunctionComponent<IPlayer> = ({ name, availableTiles }) => {
    return (
        <div className="player" title={name}>
            <span className="name" > {name} </span>
            <div className="tiles">
                {availableTiles.map(tile => (
                    <Tile key={tile.id} {...tile}  />
                ))}
            </div>      
        </div>
    );
};
