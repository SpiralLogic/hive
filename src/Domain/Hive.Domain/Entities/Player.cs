using System.Collections.Generic;

namespace Hive.Domain.Entities
{
    public sealed record Player(int Id, string Name)
    {
        public ISet<Tile> Tiles { get; init; } = new HashSet<Tile>();

        internal void RemoveTile(Tile tile)
        {
            Tiles.Remove(tile);
        }
    }
}
