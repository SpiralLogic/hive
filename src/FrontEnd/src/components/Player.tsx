import { FunctionComponent, toChildArray } from 'preact';
import { useEffect, useState } from 'preact/hooks';
import '../css/player.css';
import { PlayerId } from '../domain';
import { useClassReducer } from '../utilities/class-reducer';

type Props = {
  name: string;
  id: PlayerId;
};
const Player: FunctionComponent<Props> = (props) => {
  const { name, id } = props;
  const [classes, setClassList] = useClassReducer(`player player${id}`);
  const [hidden, setHidden] = useState(false);
  const children = toChildArray(props.children).length;
  useEffect(() => {
    if (children === 0) {
      setTimeout(() => setClassList({ type: 'add', classes: ['hide'] }), 100);
    }
  }, [children]);

  const ontransitionend = () => {
    setHidden(true);
  };

  return hidden ? null : (
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    <div class={classes} title={name} onanimationend={ontransitionend} ontransitionend={ontransitionend}>
      <div class="name">{name}</div>
      <div class="tiles">{props.children}</div>
    </div>
  );
};

Player.displayName = 'Player Tiles';
export default Player;
