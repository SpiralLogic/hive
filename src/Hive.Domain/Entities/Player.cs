using System.Collections.Generic;

namespace Hive.Domain.Entities;

public sealed record Player(int Id, string Name)
{
    public ISet<Tile> Tiles { get; init; } = new HashSet<Tile>();

    internal void RemoveTile(Tile tile)
    {
        Tiles.Remove(tile);
    }

    public bool Equals(Player? other)
    {
        if (ReferenceEquals(null, other)) return false;
        if (ReferenceEquals(this, other)) return true;
        return Id == other.Id;
    }

    public override int GetHashCode()
    {
        return Id;
    }
}
