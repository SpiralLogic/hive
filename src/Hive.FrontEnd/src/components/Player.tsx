import '../css/player.css';

import { ComponentChildren, toChildArray } from 'preact';

import { useClassSignal } from '../hooks/useClassSignal';
import { useAnimatedHide } from '../hooks/useAnimatedHide';

type Properties = {
  name: string;
  id: number;
  children?: ComponentChildren;
};

const Player = (props: Readonly<Properties>) => {
  const { name, id, children } = props;
  const [classes, classAction] = useClassSignal(`player player${id}`);
  const hasNoChildren = toChildArray(children).length === 0;
  const handlers = useAnimatedHide(hasNoChildren, [classes, classAction]);

  return (
    <section class={classes} aria-label={`${name}'s unplaced pieces`} {...handlers}>
      <h2>{name}</h2>
      <div class="tiles">{children}</div>
    </section>
  );
};

Player.displayName = 'Player Tiles';
export default Player;
