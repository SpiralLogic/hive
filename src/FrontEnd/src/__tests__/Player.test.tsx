/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { fireEvent } from '@testing-library/preact';
import { h } from 'preact';
import { mockLocation, restoreLocation } from './helpers/location';
import { renderElement } from './helpers';
import Player from '../components/Player';

describe('PlayerTiles Tests', () => {
  const playerProps = { id: 1, name: 'Player 1', hide: false, currentPlayer: 1 };

  let container: Element;
  let playerTiles: Element | null;
  describe('PlayerTile tests', () => {
    beforeEach(() => {
      global.window.history.replaceState({}, global.document.title, `/game/33/1`);
      container = renderElement(<Player {...playerProps} />);
      playerTiles = document.getElementsByClassName('player').item(0);
    });

    test('playerTiles is rendered', () => {
      expect(playerTiles).not.toBeNull();
    });

    test('is rendered with playerTiles name', () => {
      const playerName = playerTiles?.querySelector('.name');

      expect(playerName).toHaveTextContent('Player 1');
    });

    test('player is rendered with their tiles', () => {
      const tiles = playerTiles?.getElementsByClassName('tiles');
      expect(tiles).toHaveLength(1);
    });

    test('nothing is rendered with no tiles left', () => {
      const emptyTileProps = { ...playerProps, tiles: [] };
      const container = renderElement(<Player {...emptyTileProps} />);
      playerTiles = container.getElementsByClassName('playerTiles').item(1);
      expect(playerTiles).toBeNull();
    });

    test(`other keys dont navigates player`, () => {
      mockLocation({ href: '/game/33/0', pathname: '/game/33/0' });
      const playerName = renderElement(<Player {...playerProps} />).querySelector('.name');
      fireEvent.keyDown(playerName!, { key: 'Tab' });
      expect(window.location.href.endsWith('1')).not.toBe(true);
      restoreLocation();
    });
  });

  describe('PlayerTiles snapshot', () => {
    test('matches current snapshot', () => {
      expect(container).toMatchSnapshot();
    });
  });
});
