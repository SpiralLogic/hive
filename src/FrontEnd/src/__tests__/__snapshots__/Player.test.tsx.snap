// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerTiles Tests PlayerTiles snapshot matches current snapshot 1`] = `
<div
  class="player player1 hide"
  title="Player 1"
>
  <div
    class="name"
    title="share link for opponent"
  >
    Player 1
  </div>
  <div
    class="tiles"
  />
</div>
`;
